src/webgpu/api/validation/encoding/cmds/setImmediates.spec.ts

```diff
@@ -0,0 +1,209 @@
+  .paramsSubcasesOnly(u =>
+    u //
+      .combine('encoderType', kProgrammableEncoderTypes)
+      .combine('success', [true, false])
+  )
+  .fn(t => {
+    const { encoderType, success } = t.params;
+
+    function runTest(arrayBufferType: TypedArrayBufferViewConstructor) {
+      const kMinAlignmentBytes = 4;
+      const elementSize = arrayBufferType.BYTES_PER_ELEMENT;
+      const maxImmediateSize = t.device.limits.maxImmediateSize; // using this limit for tests
+      const validDataSize = success
+        ? Math.floor((maxImmediateSize - kMinAlignmentBytes) / elementSize)
+        : maxImmediateSize - kMinAlignmentBytes;
+      const validOffset = success
+        ? Math.ceil(kMinAlignmentBytes / elementSize)
+        : kMinAlignmentBytes;
+
+      const { encoder, validateFinish } = t.createEncoder(encoderType);
+      const data = new arrayBufferType(maxImmediateSize);
+
+      encoder.setImmediates(/* rangeOffset */ 0, data, /* dataOffset */ validOffset, validDataSize);
+
+      validateFinish(success || elementSize === 1);
+    }
+
+    for (const arrayType of kTypedArrayBufferViewConstructors) {
+      if (arrayType === Float16Array) {
+        // Skip Float16Array since it is supplied by an external module, so there isn't an overload for it.
+        continue;
+      }
+      runTest(arrayType);
+    }
+  });
+
+g.test('alignment')
+  .desc('Tests that rangeOffset and contentSize must align to 4 bytes.')
+  .paramsSubcasesOnly(u =>
+    u //
+      .combine('encoderType', kProgrammableEncoderTypes)
+      .combineWithParams([
+        // control case
+        { offset: 4, contentByteSize: 4, _offsetValid: true, _contentValid: true },
+        // offset is not aligned to 4 bytes
+        { offset: 1, contentByteSize: 4, _offsetValid: false, _contentValid: true },
+        // contentSize is not aligned to 4 bytes
+        { offset: 4, contentByteSize: 5, _offsetValid: true, _contentValid: false },
+      ] as const)
+  )
+  .fn(t => {
+    const { encoderType, offset, contentByteSize, _offsetValid, _contentValid } = t.params;
+    const data = new Uint8Array(contentByteSize);
+
+    const { encoder, validateFinish } = t.createEncoder(encoderType);
+
+    const doSetImmediates = () => {
+      encoder.setImmediates(offset, data, 0, contentByteSize);
+    };
+
+    if (_contentValid) {
+      doSetImmediates();
+    } else {
+      t.shouldThrow('RangeError', doSetImmediates);
+    }
+
+    validateFinish(_offsetValid);
+  });
+
+g.test('overflow')
+  .desc('Tests that rangeOffset + contentSize exceeds Number.MAX_SAFE_INTEGER.')
+  .paramsSubcasesOnly(u =>
+    u //
+      .combine('encoderType', kProgrammableEncoderTypes)
+      .combineWithParams([
+        // control case
+        { offset: 4, contentByteSize: 4, _rangeValid: true, _contentValid: true },
+        // rangeOffset + contentSize overflows
+        {
+          offset: 8,
+          contentByteSize: kMaxSafeMultipleOf8,
```

Comment:
I think this should test `2**31 - 8` and `2**32 - 8`. That's closer to "reasonable" so it's more likely to find bugs before being caught by some hardcoded OOM limit.

I can't think of a bug that would be caught by `2**53 - 8` but not by `2**32 - 8`. And of course we can't test `2**64 - 8` since it's a `double`.